src/tactic/induction.lean

```diff
@@ -495,10 +496,61 @@ meta def constructor_intros (generate_induction_hyps : bool)
 meta def ih_name (arg_name : name) : name :=
 mk_simple_name ("ih_" ++ arg_name.to_string)
 
-private meta def get_with_name : option name → option name
-| (some `_) := none
-| (some n) := some n
-| none := none
+/--
+Representation of a pattern in the `with n ...` syntax supported by
+`induction'` and `cases'`. We support the following patterns:
+
+- `auto` (`with _` or no `with` clause): use the name generated by the tactic.
+- `clear` (`with -`): clear this hypothesis and any hypotheses depending on it.
+- `exact n` (`with n`): use the name `n` for this hypothesis.
+-/
+@[derive has_reflect]
+meta inductive with_pattern
+| auto
+| clear
+| exact (n : name)
+
+namespace with_pattern
+
+open lean
+open lean.parser
+
+/-- Parser for a `with_pattern`. -/
+meta def parser : lean.parser with_pattern :=
+(tk "-" *> pure with_pattern.clear) <|>
+(tk "_" *> pure with_pattern.auto) <|>
+(with_pattern.exact <$> ident)
+
+/-- Parser for a `with` clause. -/
+meta def clause_parser : lean.parser (list with_pattern) :=
+(tk "with" *> many parser) <|> pure []
```

Comment:
The definition shows that `parser` can lead to naming conflicts with `lean.parser`, so I would make it a `protected def` instead.
```suggestion
meta protected def parser : parser with_pattern :=
(tk "-" *> pure with_pattern.clear) <|>
(tk "_" *> pure with_pattern.auto) <|>
(with_pattern.exact <$> ident)

/-- Parser for a `with` clause. -/
meta def clause_parser : parser (list with_pattern) :=
(tk "with" *> many with_pattern.parser) <|> pure []
```